### PR TITLE
CLOUDP-222194: Manager Builder

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -72,219 +48,182 @@ const (
...
+	mgr, err := operator.NewBuilder().
+		WithNamespaces(collection.MapKeysToSlice(config.WatchedNamespaces)...).
+		WithLogger(logger).
+		WithMetricAddress(config.MetricsAddr).
+		WithProbeAddress(config.ProbeAddr).
+		WithLeaderElection(config.EnableLeaderElection).
+		WithAtlasDomain(config.AtlasDomain).
+		WithAPISecret(config.GlobalAPISecret).
+		WithDeletionProtection(config.ObjectDeletionProtection).
+		Build(ctx)
```
[the remark this answers is not in the document]
Comment (josvazg) [reply cmd/manager/main.go:74]: On second though, there is really not much of a difference in safety between this builder and the struct. And we do need different initalization specially in integration tests. It is sufficient that we keep e2e tests in sync with main as much as possible.

So yes, let's go the builder route.

File: cmd/manager/main.go [hunk 2]
```diff
@@ -72,219 +48,182 @@ const (
...
 
-var (
-	scheme   = runtime.NewScheme()
-	setupLog = ctrl.Log.WithName("setup")
-)
-
-func init() {
-	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
-	utilruntime.Must(akov2.AddToScheme(scheme))
-}
```
Comment (s-urbaniak) [inline cmd/manager/main.go:83]: very nice removing this from package level initialization 👏 🎉

File: cmd/manager/main.go [hunk 3]
```diff
@@ -72,219 +48,182 @@ const (
...
-
-	if err := mgr.AddHealthzCheck("health", healthz.Ping); err != nil {
-		setupLog.Error(err, "unable to set up health check")
-		os.Exit(1)
-	}
-	if err := mgr.AddReadyzCheck("check", healthz.Ping); err != nil {
-		setupLog.Error(err, "unable to set up ready check")
-		os.Exit(1)
-	}
+	/*
```
Comment (s-urbaniak) [inline cmd/manager/main.go:83]: remove commented code

File: pkg/controller/watch/predicates.go [hunk 4]
```diff
@@ -21,29 +21,16 @@ func CommonPredicates() predicate.Funcs {
...
-			return false
-		},
-		GenericFunc: func(ce event.GenericEvent) bool {
-			return false
-		},
-	}
-}
-
-func SelectNamespacesPredicate(namespaceMap map[string]bool) predicate.Funcs {
+func SelectNamespacesPredicate(namespaces []string) predicate.Funcs {
```
Comment (s-urbaniak) [inline pkg/controller/watch/predicates.go:24]: please add a unit test for this function

File: pkg/indexer/atlasstreaminstances.go [hunk 5]
```diff
@@ -48,3 +50,37 @@ func (a *AtlasStreamInstanceByConnectionIndexer) Keys(object client.Object) []st
 
 	return indices
 }
+
+type AtlasStreamInstanceByProjectIndexer struct {
+	logger *zap.SugaredLogger
+}
+
+func NewAtlasStreamInstanceByProjectIndexer(logger *zap.Logger) *AtlasStreamInstanceByProjectIndexer {
```
Comment (s-urbaniak) [inline pkg/indexer/atlasstreaminstances.go:58]: unrelated to the builder?

File: pkg/operator/builder.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+
+	if err = indexer.RegisterAll(ctx, mgr, b.logger); err != nil {
+		return nil, fmt.Errorf("unable to create indexers: %w", err)
+	}
+
+	if b.atlasProvider == nil {
+		b.atlasProvider = atlas.NewProductionProvider(b.atlasDomain, b.apiSecret, mgr.GetClient())
+	}
+
+	if err = (&atlasproject.AtlasProjectReconciler{
```
Comment (s-urbaniak) [inline pkg/operator/builder.go:198]: since we're here. one thing that I recognized for quite some time is that none of the reconciler structs has `New...` constructor methods. Please add `NewAtlasProjectReconciler(...) *atlasproject.AtlasProjectReconciler` for this and the reconcilers below.

File: pkg/operator/builder.go [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+	mergeDefaults(b)
+
+	utilruntime.Must(clientgoscheme.AddToScheme(b.scheme))
+	utilruntime.Must(akov2.AddToScheme(b.scheme))
+
+	cacheOpts := cache.Options{
+		SyncPeriod: &b.syncPeriod,
+	}
+
+	switch len(b.namespaces) {
```
Comment (s-urbaniak) [inline pkg/operator/builder.go:148]: i'm a big fan of switch statements but I guess a `if len(b.namespaces) == 0 { ... } else { ... }` is more readable here.

File: pkg/operator/builder.go [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+	case 0:
+		cacheOpts.ByObject = map[client.Object]cache.ByObject{
+			&corev1.Secret{}: {
```
Comment (s-urbaniak) [inline pkg/operator/builder.go:151]: why only Secret?

File: pkg/operator/builder.go [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+		AtlasProvider:            b.atlasProvider,
+		ObjectDeletionProtection: b.deletionProtection,
+	}).SetupWithManager(ctx, mgr); err != nil {
+		return nil, fmt.Errorf("unable to create controller AtlasSearchIndexConfig: %w", err)
+	}
+
+	return mgr, nil
+}
+
+func NewBuilder() *Builder {
```
Comment (s-urbaniak) [inline pkg/operator/builder.go:302]: please add godoc for all public methods

File: pkg/operator/builder.go [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+
+const (
+	DefaultAtlasDomain      = "https://cloud.mongodb.com/"
+	DefaultSyncPeriod       = 3 * time.Hour
+	DefaultMetricAddress    = ":8080"
+	DefaultProbeAddress     = ":8081"
+	DefaultLeaderElectionID = "06d035fb.mongodb.com"
+)
+
+type Builder struct {
```
Comment (s-urbaniak) [inline pkg/operator/builder.go:49]: pleas add unit tests for the builder package

**Comment by s-urbaniak** [comment]
This goes in a good direction, I like it 👍

**Comment by s-urbaniak** [comment]
whops, sorry, accidentally closed this one 🙈 (and reopened)

File: pkg/operator/builder.go (hunk 8 above)
Comment (s-urbaniak) [reply pkg/operator/builder.go:151]: TIL, it is to limit listening to secrets in the whole cluster. we are only interested in secrets marked as "AKO secrets" using a label

File: pkg/operator/builder.go [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+func (b *Builder) WithDeletionProtection(deletionProtection bool) *Builder {
+	b.deletionProtection = deletionProtection
+	return b
+}
+
+func (b *Builder) Build(ctx context.Context) (manager.Manager, error) {
+	mergeDefaults(b)
+
+	utilruntime.Must(clientgoscheme.AddToScheme(b.scheme))
+	utilruntime.Must(akov2.AddToScheme(b.scheme))
```
Comment (s-urbaniak) [inline pkg/operator/builder.go:142]: i wouldn't do this here

File: internal/collection/collection.go [hunk 12]
```diff
@@ -11,3 +11,13 @@ func CopyWithSkip[T comparable](list []T, skip T) []T {
 
 	return newList
 }
+
+func MapKeysToSlice[K comparable, V any](m map[K]V) []K {
```
Comment (josvazg) [inline internal/collection/collection.go:15]: This could be simply called `Keys`:
```suggestion
func Keys[K comparable, V any](m map[K]V) []K {
```

File: pkg/controller/watch/predicates.go (hunk 4 above)
Comment (josvazg) [reply pkg/controller/watch/predicates.go:24]: +1

File: pkg/controller/watch/predicates_test.go [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+)
+
+func TestSelectNamespacesPredicate(t *testing.T) {
```
Comment (josvazg) [inline pkg/controller/watch/predicates_test.go:13]: Nit: This would fit a table driven test perfectly.

File: pkg/operator/builder.go [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+	config         *rest.Config
+	namespaces     []string
+	logger         *zap.Logger
+	syncPeriod     time.Duration
+	metricAddress  string
+	probeAddress   string
+	leaderElection bool
+
+	leaderElectionID   string
+	atlasDomain        string
```
Comment (josvazg) [inline pkg/operator/builder.go:67]: Why separate these?

```suggestion
	probeAddress     string
	leaderElection   bool
	leaderElectionID string
	
	atlasDomain        string
```

File: pkg/operator/builder.go [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+
+	if err = indexer.RegisterAll(ctx, mgr, b.logger); err != nil {
+		return nil, fmt.Errorf("unable to create indexers: %w", err)
+	}
+
+	if b.atlasProvider == nil {
+		b.atlasProvider = atlas.NewProductionProvider(b.atlasDomain, b.apiSecret, mgr.GetClient())
+	}
+
+	_, err = atlasproject.NewAtlasProjectReconciler(
```
Comment (josvazg) [inline pkg/operator/builder.go:198]: We are ignoring the reconciler everywhere, why not instead this?

```suggestion
	err = atlasproject.RegisterAtlasProjectReconciler(
```

It is basically registering the reconciler within the manager right?

File: test/e2e/db_users_oidc_test.go [hunk 16]
```diff
@@ -73,8 +73,8 @@ var _ = Describe("Operator to run db-user with the OIDC feature flags", Ordered,
 					Namespace: config.DefaultOperatorNS,
 					Name:      config.DefaultOperatorGlobalKey,
 				},
-				Namespace:    config.DefaultOperatorNS,
-				FeatureFlags: featureflags.NewFeatureFlags(func() []string { return []string{} }),
+				WatchedNamespaces: map[string]bool{config.DefaultOperatorNS: true},
```
Comment (josvazg) [inline test/e2e/db_users_oidc_test.go:76]: I believe in my watch namespaces fix, this is gone. But it is not important, we will merge sooner or later.

File: test/helper/e2e/k8s/operator.go [hunk 17]
```diff
@@ -47,162 +30,43 @@ var (
 )
 
 func BuildManager(initCfg *Config) (manager.Manager, error) {
-	scheme := runtime.NewScheme()
-	setupLog := ctrl.Log.WithName("setup")
+	akoScheme := runtime.NewScheme()
+	utilruntime.Must(scheme.AddToScheme(akoScheme))
+	utilruntime.Must(akov2.AddToScheme(akoScheme))
```
Comment (josvazg) [inline test/helper/e2e/k8s/operator.go:35]: We will have to move this away from here. But we do not need to do it here.
See CLOUDP-251767 but basically we want this add to scheme to be done once at the test suite setup. Otherwise you cannot launch several managers with different configurations, as they will race with each other.

File: test/int/clusterwide/dbuser_test.go [hunk 18]
```diff
@@ -137,6 +138,9 @@ func buildConnectionSecret(name string) corev1.Secret {
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      name,
 			Namespace: namespace.Name,
+			Labels: map[string]string{
+				connectionsecret.TypeLabelKey: connectionsecret.CredLabelVal,
+			},
```
Comment (josvazg) [inline test/int/clusterwide/dbuser_test.go:143]: Why was this addition needed?

**Review by josvazg — APPROVED** [review]
LGTM, with some nits and questions

File: pkg/operator/builder.go (hunk 15 above)
Comment (helderjs) [reply pkg/operator/builder.go:198]: it does construct the reconciler, register it tho the manager is a bonus(if we can call it like this)

File: test/e2e/db_users_oidc_test.go (hunk 16 above)
Comment (helderjs) [reply test/e2e/db_users_oidc_test.go:76]: Not really, this is a namespaced test

File: test/helper/e2e/k8s/operator.go (hunk 17 above)
Comment (helderjs) [reply test/helper/e2e/k8s/operator.go:35]: I actually gave a try, but everything has fallen apart and I decided to solve this in a separate PR

File: test/int/clusterwide/dbuser_test.go (hunk 18 above)
Comment (helderjs) [reply test/int/clusterwide/dbuser_test.go:143]: The manager on the cluster-wide integration test was built very differently from the production version. Being cluster-wide, we should be watching only labeled secrets, and with the fix, this had to be added.

File: pkg/operator/builder.go (hunk 15 above)
Comment (josvazg) [reply pkg/operator/builder.go:198]: Yes but my point is that the main usage of these functions do NOT use the returned reconcilers so, why do we return it at all in the signature?

I mean there seems to be a API impedance mismatch here.

File: pkg/indexer/atlasstreaminstances.go (hunk 5 above)
Comment (helderjs) [reply pkg/indexer/atlasstreaminstances.go]: The PR was requiring rebase

File: pkg/controller/atlasproject/atlasproject_controller.go [hunk 19]
```diff
@@ -234,47 +234,58 @@ func (r *AtlasProjectReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	return workflow.OK().ReconcileResult(), nil
 }
 
-func (r *AtlasProjectReconciler) handleDeletion(workflowCtx *workflow.Context, atlasClient *mongodbatlas.Client, project *akov2.AtlasProject) (result workflow.Result) {
-	log := workflowCtx.Log
+func (r *AtlasProjectReconciler) handleDeletion(ctx *workflow.Context, project *akov2.AtlasProject) (result workflow.Result) {
```
[the remark this answers is not in the document]
Comment (helderjs) [reply pkg/controller/atlasproject/atlasproject_controller.go]: The PR was requiring rebase

File: cmd/manager/main.go [hunk 20]
```diff
@@ -308,7 +112,7 @@ type Config struct {
 func parseConfiguration() Config {
 	var globalAPISecretName string
 	config := Config{}
-	flag.StringVar(&config.AtlasDomain, "atlas-domain", "https://cloud.mongodb.com/", "the Atlas URL domain name (with slash in the end).")
+	flag.StringVar(&config.AtlasDomain, "atlas-domain", operator.DefaultAtlasDomain, "the Atlas URL domain name (with slash in the end).")
```
Comment (josvazg) [inline cmd/manager/main.go:115]: > "(with slash in the end)."

Unrelated for this PR but should we make it so that this comment is not needed and the `/` is auto auto appended as needed?

Comment (helderjs) [reply cmd/manager/main.go:115]: Why is this a problem?

**Review by josvazg — APPROVED** [review]
Still LGTM, provided CI agrees as well

File: cmd/manager/main.go (hunk 20 above)
Comment (josvazg) [reply cmd/manager/main.go:115]: Maybe not important, but if you forget the `/` in the helm chart values you will not have this nice help message available to you there. I mean, usually the less explanations needed the better.
But is off topic, not for this PR.
